libbpf-tools/profile.c

```diff
@@ -268,13 +270,61 @@ static int cmp_counts(const void *a, const void *b)
 	return y - x;
 }
 
+static int read_batch_counts_map(int fd, struct key_ext_t *items, __u32 *count)
+{
+	struct key_t keys[*count];
+	__u64 vals[*count];
+	void *in = NULL, *out;
+	__u32 i, n, n_read = 0;
+	int err = 0;
+
+	while (n_read < *count && !err) {
+		n = *count - n_read;
+		err = bpf_map_lookup_batch(fd, &in, &out, keys + n_read,
+					   vals + n_read, &n, NULL);
+		if (err < 0 && err != -ENOENT) {
+			/* we want to propagate EINVAL upper, so that
+			 * the batch_map_ops flag is set to false */
+			if (err != -EINVAL)
+				fprintf(stderr, "bpf_map_lookup_batch: %s\n",
+					strerror(-err));
+
+			return err;
+		}
+
+		n_read += n;
+		in = out;
+	}
+
+	for (i = 0; i < n_read; i++) {
+		items[i].k.pid = keys[i].pid;
+		items[i].k.user_stack_id = keys[i].user_stack_id;
+		items[i].k.kern_stack_id = keys[i].kern_stack_id;
+		strncpy(items[i].k.name, keys[i].name, TASK_COMM_LEN);
+		items[i].v = vals[i];
+	}
+
+	*count = n_read;
+	return 0;
+}
+
 static int read_counts_map(int fd, struct key_ext_t *items, __u32 *count)
 {
 	struct key_t empty = {};
 	struct key_t *lookup_key = &empty;
 	int i = 0;
 	int err;
 
+	if (batch_map_ops) {
+		err = read_batch_counts_map(fd, items, count);
+		if (err < 0 && err == -EINVAL) {
+			/* fall back to a racy variant */
+			batch_map_ops = false;
+		} else {
+			return err;
+		}
+	}
```

Comment:
In map_helpers.c, we have dump_hash(). Could you try to use that function?

Reply:
Yes, there is duplicate code in this patch.
read_counts_map() in profile.c returns data as an array of type "struct key_ext_t", which is a combination of the keys and values ​​from the hash map. This is the difference from the design of dump_hash() in map_helpers.c.

In the current patch, the combination is only needed for lookup_batch. However, using map_helpers' dump_hash() requires combination (data copying) for both the batch and iteration cases. Perhaps that's why syscount.c doesn't use map_helpers either. Could you tell me what you think about it?

There is also a way to utilize only dump_hash_batch() of map_helpers.c after removing static.
Thank you.

Reply:
I think dump_hash() should work. It will cover both batch and non-batch case. I suggest to replace non-batch case with dump_hash(). Could you give a try?

Reply:
Of course, I'll update it.